### PR TITLE
intial RPO loss

### PR DESCRIPTION
**Pull request by kashif**
from the paper: https://arxiv.org/pdf/2404.19733

### CODE REVIEW

**Review by younesbelkada — APPROVED** [review]
Thanks !

**Comment by winglian** [comment]
@kashif couldn't get a clean stack trace on multi-gpu/process, but doesn't seem to work with deepspeed:

```
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/trainer.py", line 1885, in train                                                                                 
    return inner_training_loop(                                                                                                                                                                 
  Fil  "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/trainer.py", line 2216, in _inner_training_loop                                                                  
    tr_loss_step = self.training_step(model, inputs)                                                                                                                                            
    self.optimizer.step()                                                                                                                                                                       
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/trainer.py", line 3250, in training_step                                                                         
    self.accelerator.backward(loss)                                                                                                                                                             
  Fil  "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1846, in step                                                                  
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/accelerator.py", line 2117, in backward                                                                            
    self.deepspeed_engine_wrapped.backward(loss, **kwargs)                                                                                                                                      
  Fil  "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/utils/deepspeed.py", line 175, in backward                                                                         
    self.engine.step()                                                                                                                                                                          
    r turn torch.norm(torch.stack(norm_groups), p=norm_type)                                                                                                                          [342/1890]
self.optimizer.step()  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/engine.py", line 2169, in step                                                         
    self._take_model_step(lr_kwargs)                                                                                                                                                            
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/functional.py", line 1611, in norm                                                                                      
                                                                                                                                                                                                
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1846, in step                                                                  
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/engine.py", line 2075, in _take_model_step                                                                  
    self.optimizer.step()                                                                                                                                                                       
         root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1846, in step                                                                  
    scaled_global_grad_norm = self.scaled_global_norm()                                                                                                                                         
          oot/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1793, in scaled_global_norm                                                    
    return torch.norm(torch.stack(norm_groups), p=norm_type)                                                                                                                                    
    scaled_global_grad_norm = self.scaled_global_norm()  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/functional.py", line 1611, in norm                               
    return torch.linalg.vector_norm(input, _p, _dim, keepdim, dtype=dtype)                                                                                                                      
                                                                                                                                                                                                
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1793, in scaled_global_norm                                                    
RuntimeError: linalg.vector_norm: Expected a floating point or complex tensor as input. Got Long                                                                                                
    return torch.linalg.vector_norm(input, _p, _dim, keepdim, dtype=dtype)                                                                                                                      
RuntimeError: linalg.vector_norm: Expected a floating point or complex tensor as input. Got Long                                                                                                
    scaled_global_grad_norm = self.scaled_global_norm()                                                                                                                                         
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1793, in scaled_global_norm                                                        return torch.norm(torch.stack(norm_groups), p=norm_type)                                                                                                                                    
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/functional.py", line 1611, in norm                                                                                      
    return torch.norm(torch.stack(norm_groups), p=norm_type)                                                                                                                                    
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/functional.py", line 1611, in norm                                                                                      
    return torch.linalg.vector_norm(input, _p, _dim, keepdim, dtype=dtype)                                                                                                                      
RuntimeError: linalg.vector_norm: Expected a floating point or complex tensor as input. Got Long                                                                                                
    return torch.linalg.vector_norm(input, _p, _dim, keepdim, dtype=dtype)                                                                                                                      
RuntimeError: linalg.vector_norm: Expected a floating point or complex tensor as input. Got Long  
```

**Comment by kashif** [comment]
thanks @winglian  checking!

**Comment by kashif** [comment]
@winglian are you still encountering this issue?

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1184,10 +1185,15 @@ def get_batch_loss_metrics(
...
             policy_rejected_logits,
+            policy_chosen_logps_avg,
         ) = self.concatenated_forward(model, batch)
 
         # if reference_chosen_logps and reference_rejected_logps in batch use them, otherwise use the reference model
-        if "reference_chosen_logps" in batch and "reference_rejected_logps" in batch:
+        if (
+            "reference_chosen_logps" in batch
+            and "reference_rejected_logps" in batch
+            and self.args.rpo_alpha is not None
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1195]: why this condition?

**Pull request merged**